cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/oci_repo.go

```diff
@@ -0,0 +1,645 @@
+	enc := gob.NewEncoder(&buf)
+	if err = enc.Encode(cacheEntryValue); err != nil {
+		return nil, false, err
+	}
+	return buf.Bytes(), true, nil
+}
+
+func (s *repoEventSink) onModifyOciRepo(key string, oldValue interface{}, repo sourcev1.HelmRepository) ([]byte, bool, error) {
+	log.Infof("+onModifyOciRepo(%s)", common.PrettyPrint(repo))
+	defer log.Infof("-onModifyOciRepo")
+
+	// We should to compare checksums on what's stored in the cache
+	// vs the modified object to see if the contents has really changed before embarking on
+	// an expensive operation
+	cacheEntryUntyped, err := s.onGetRepo(key, oldValue)
+	if err != nil {
+		return nil, false, err
+	}
+
+	cacheEntry, ok := cacheEntryUntyped.(repoCacheEntryValue)
+	if !ok {
+		return nil, false, status.Errorf(
+			codes.Internal,
+			"unexpected value found in cache for key [%s]: %v",
+			key, cacheEntryUntyped)
+	}
+
+	ociRegistry, err := s.newOCIRegistryAndLogin(repo)
+	if err != nil {
+		return nil, false, err
+	}
+
+	newChecksum, err := ociRegistry.checksum()
+	if err != nil {
+		return nil, false, err
+	}
+
+	if cacheEntry.Checksum != newChecksum {
+		return nil, false, status.Errorf(codes.Unimplemented, "TODO")
+	} else {
+		// skip because the content did not change
+		return nil, false, nil
+	}
+}
+
+//
+// misc OCI repo utilities
+//
+
+// TagList represents a list of tags as specified at
+// https://github.com/opencontainers/distribution-spec/blob/main/spec.md#content-discovery
+type TagList struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+// Checksum returns the sha256 of the repo by concatenating tags for
+// all repositories within the registry and returning the sha256.
+// Caveat: Mutated image tags won't be detected as new
+func (r *OCIRegistry) checksum() (string, error) {
+	appNames, err := r.listRepositoryNames()
+	if err != nil {
+		return "", err
+	}
+	tags := map[string]TagList{}
+	for _, appName := range appNames {
+		ref := fmt.Sprintf("%s/%s", r.url.String(), appName)
+		tagss, err := r.getTags(ref)
+		if err != nil {
+			return "", err
+		}
+		tags[appName] = TagList{Name: appName, Tags: tagss}
+	}
+
+	content, err := json.Marshal(tags)
+	if err != nil {
+		return "", err
+	}
+
+	return common.GetSha256(content)
+}
+
+func (s *repoEventSink) newOCIRegistryAndLogin(repo sourcev1.HelmRepository) (*OCIRegistry, error) {
+	// Construct the Getter options from the HelmRepository data
+	loginOpts, getterOpts, cred, err := s.clientOptionsForRepo(repo)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create registry client options: %v", err)
+	}
+	log.Infof("=====================> loginOpts: [%v], getterOpts: [%v], cred: %t", len(loginOpts), len(getterOpts), cred != nil)
+
+	// Create registry client and login if needed.
+	registryClient, file, err := newRegistryClient(loginOpts != nil)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create registry client: %v", err)
+	}
+	if file != "" {
+		/*
```

Comment:
Looks like there's a bit of debugging logging and commented out code here - not sure if you wanted it reviewed yet in its current state? I'm ignoring the commented out stuff, but wondering why it's here.

Reply:
work in progress. No need to review commented out code, of course. These are just notes to myself. It'll be clean at the end, I pinky promise